File: 프로그래머스/3/76503. 모두 0으로 만들기/모두 0으로 만들기.py
import sys
import math

sys.setrecursionlimit(300000)

def solution(a, edges):
    total = sum(a)
    if total != 0:
        return -1
    
    tree = dict()
    for [u,v] in edges:
        if u not in tree: 
            tree[u] = []
        if v not in tree: 
            tree[v] = []
        
        tree[u].append(v)
        tree[v].append(u)
    
    answer = 0
    visited = set()
    
    def dfs(node, parent):
        nonlocal answer
        
        visited.add(node)
        node_sum = a[node]
        
        for next in tree[node]:
            if next in visited or next == parent:
                continue
            
            child_sum = dfs(next, node)
            node_sum += child_sum
            answer += abs(child_sum)
        
        return node_sum
    
    dfs(0, -1)
    return answer
    

File: 프로그래머스/3/76503. 모두 0으로 만들기/모두 0으로 만들기.js
function solution(a, edges) {
    let sum = a.reduce((sum, cur)=> sum+cur, 0)
    if(sum !== 0) return -1;
    
    const tree = {};
    
    for(const [u,v] of edges){
        tree[u] ? tree[u].push(v) : tree[u] = [v]
        tree[v] ? tree[v].push(u) : tree[v] = [u]
    }

    const visited = new Set();
    let answer = 0n;
    
    const stack = [[0, null]]; // [node, parent]
    const sums = Array(a.length).fill(0n);
   
    while(stack.length) {
       const [node, parent] = stack[stack.length - 1];
       
       if(!visited.has(node)) { // 방문한 적 없으면
           visited.add(node);
           sums[node] = BigInt(a[node]);
           
           for(const next of tree[node]) { // 방문하고 부모 아닌 다음 위치로 이동
               if(next !== parent) stack.push([next, node]);
           }
       } else {
           stack.pop(); // 방문 했으면 탐색 X
           if(parent !== null) { 
               sums[parent] += sums[node];
               answer += BigInt(Math.abs(Number(sums[node])));
           }
       }
   }
    
   return answer;
}